src/analysis/processing/qgsalgorithmbatchgeocode.cpp

```diff
@@ -0,0 +1,194 @@
+  }
+}
+
+QList<int> QgsBatchGeocodeAlgorithm::inputLayerTypes() const
+{
+  return QList<int>() << QgsProcessing::TypeVector;
+}
+
+bool QgsBatchGeocodeAlgorithm::supportInPlaceEdit( const QgsMapLayer *layer ) const
+{
+  if ( const QgsVectorLayer *vl = qobject_cast< const QgsVectorLayer * >( layer ) )
+  {
+    return vl->geometryType() == QgsWkbTypes::PointGeometry;
+  }
+  return false;
+}
+
+QString QgsBatchGeocodeAlgorithm::outputName() const
+{
+  return QObject::tr( "Geocoded" );
+}
+
+bool QgsBatchGeocodeAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  mAddressField = parameterAsString( parameters, QStringLiteral( "FIELD" ), context );
+
+  if ( mIsInPlace )
+  {
+    const QgsFields newFields = mGeocoder->appendedFields();
+    for ( const QgsField &newField : newFields )
+      mInPlaceFieldMap.insert( newField.name(), parameterAsString( parameters, newField.name(), context ) );
+  }
+
+  return true;
+}
+
+QgsWkbTypes::Type QgsBatchGeocodeAlgorithm::outputWkbType( QgsWkbTypes::Type ) const
+{
+  return QgsWkbTypes::Point;
+}
+
+QgsCoordinateReferenceSystem QgsBatchGeocodeAlgorithm::outputCrs( const QgsCoordinateReferenceSystem &inputCrs ) const
+{
+  mOutputCrs = inputCrs;
+  return mOutputCrs;
+}
+
+QgsFields QgsBatchGeocodeAlgorithm::outputFields( const QgsFields &inputFields ) const
+{
+  if ( !mIsInPlace )
+  {
+    // append any additional fields created by the geocoder
+    const QgsFields newFields = mGeocoder->appendedFields();
+    mAdditionalFields = newFields.names();
+
+    return QgsProcessingUtils::combineFields( inputFields, newFields );
+  }
+  else
+  {
+    return inputFields;
+  }
+}
+
+QgsFeatureList QgsBatchGeocodeAlgorithm::processFeature( const QgsFeature &feature, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  QgsFeature f = feature;
+
+  QgsAttributes attr = f.attributes();
+
+  const QString address = f.attribute( mAddressField ).toString();
+  if ( address.isEmpty() )
+  {
+    // append null attributes
+    for ( int i = 0; i < mAdditionalFields.count(); ++ i )
+      attr.append( QVariant() );
+    f.setAttributes( attr );
+    feedback->reportError( QObject::tr( "Empty address field for feature %1" ).arg( feature.id() ) );
+    return QgsFeatureList() << f;
+  }
+
+  QgsGeocoderContext geocodeContext( context.transformContext() );
+  const QList< QgsGeocoderResult > results = mGeocoder->geocodeString( address, geocodeContext, feedback );
+  if ( results.empty() )
+  {
+    // append null attributes
+    for ( int i = 0; i < mAdditionalFields.count(); ++ i )
+      attr.append( QVariant() );
+    f.setAttributes( attr );
+    feedback->reportError( QObject::tr( "No result for %1" ).arg( address ) );
+    return QgsFeatureList() << f;
+  }
+
+  if ( !results.at( 0 ).isValid() )
```

Comment:
What about merging this conditional block with the one above (i.e. the if (  results.empty() ). The only difference is the warning message.

Reply:
Actually it's a little different -- if a result has !isValid() then it indicates that it's an error result. From the dox:

```
    /**
     * Returns TRUE if the result is a valid result.
     *
     * If the result is invalid, the error message can be retrieved by calling error().
     */
```

So in this case it's catching a special case where an error was returned by the service, not no results.

Reply:
Right, I was thinking you could do 

```
if ( results.isEmpty() || !results.at( 0 ).isValid() {
...
de-duplicated code
...
feedback->reportError( results.isEmpty() ? QObject::tr( "No result for %1" ).arg( address ) : QObject::tr( "Error geocoding %1: %2" ).arg( address, results.at( 0 ).error() ) );
}
```

Reply:
But then we're mixing an error and warning result :)

Reply:
(ie, we'd also have to have an if statement to either call reportError or pushWarning respectively...)

Reply:
Okay :)

Reply:
Let me do a compromise and fix the ultimate underlying issue here...

Reply:
Does https://github.com/qgis/QGIS/pull/40339/commits/821c81a44ccc25cf92c0c821f1498b5561cde212 satisfy you?

Reply:
@nyalldawson , nice. It'll come in handy in many places. Thanks